drivers/gpio/gpio_sam.c

```diff
@@ -16,78 +16,99 @@
 };
 
 #define DEV_CFG(dev) \
-	((const struct gpio_sam_config *const)(dev)->config->config_info)
+	((const struct gpio_sam_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct gpio_sam_runtime * const)(dev)->driver_data)
 
-static int gpio_sam_config_pin(Pio * const pio, u32_t mask, int flags)
+#define GPIO_SAM_ALL_PINS    0xFFFFFFFF
+
+static int gpio_sam_port_configure(struct device *dev, u32_t mask,
+				   gpio_flags_t flags)
 {
-	/* Setup the pin direcion. */
-	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
-		pio->PIO_OER = mask;
-	} else {
-		pio->PIO_ODR = mask;
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Pio * const pio = cfg->regs;
+
+	if (flags & GPIO_SINGLE_ENDED) {
+		/* TODO: Add support for Open Source, Open Drain mode */
+		return -ENOTSUP;
 	}
 
-	/* Setup interrupt configuration. */
-	if (flags & GPIO_INT) {
-		if (flags & GPIO_INT_DOUBLE_EDGE) {
-			return -ENOTSUP;
-		}
+	if (!(flags & (GPIO_OUTPUT | GPIO_INPUT))) {
+		/* Neither input nor output mode is selected */
 
-		/* Enable the interrupt. */
-		pio->PIO_IER = mask;
+		/* Disable the interrupt. */
+		pio->PIO_IDR = mask;
+		/* Disable pull-up. */
+		pio->PIO_PUDR = mask;
+#if defined(CONFIG_SOC_SERIES_SAM4S) || defined(CONFIG_SOC_SERIES_SAME70)
+		/* Disable pull-down. */
+		pio->PIO_PPDDR = mask;
+#endif
+		/* Let the PIO control the pin (instead of a peripheral). */
+		pio->PIO_PER = mask;
+		/* Disable output. */
+		pio->PIO_ODR = mask;
 
-		/* Enable the additional interrupt modes. */
-		pio->PIO_AIMER = mask;
+		return 0;
+	}
 
-		if (flags & GPIO_INT_EDGE) {
-			pio->PIO_ESR = mask;
-		} else {
-			pio->PIO_LSR = mask;
+	/* Setup the pin direcion. */
+	if (flags & GPIO_OUTPUT) {
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			/* Set the pin. */
+			pio->PIO_SODR = mask;
 		}
-
-		if (flags & GPIO_INT_ACTIVE_HIGH) {
-			/* Set to high-level or rising edge. */
-			pio->PIO_REHLSR = mask;
-		} else {
-			/* Set to low-level or falling edge. */
-			pio->PIO_FELLSR = mask;
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			/* Clear the pin. */
+			pio->PIO_CODR = mask;
 		}
+		/* Enable the output */
+		pio->PIO_OER = mask;
+		/* Enable direct control of output level via PIO_ODSR */
+		pio->PIO_OWER = mask;
 	} else {
-		/* Disable the interrupt. */
-		pio->PIO_IDR = mask;
-
-		/* Disable their additional interrupt modes. */
-		pio->PIO_AIMDR = mask;
+		/* Disable the output */
+		pio->PIO_ODR = mask;
 	}
 
-	/* Setup Pull-up resistor. */
-	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+	/* Note: Input is always enabled. */
+
+	/* Setup selected Pull resistor.
+	 *
+	 * A pull cannot be enabled if the opposite pull is enabled.
+	 * Clear both pulls, then enable the one we need.
+	 */
+	pio->PIO_PUDR = mask;
+#if defined(CONFIG_SOC_SERIES_SAM4S) || defined(CONFIG_SOC_SERIES_SAME70)
+	pio->PIO_PPDDR = mask;
+#endif
+	if (flags & GPIO_PULL_UP) {
 		/* Enable pull-up. */
 		pio->PIO_PUER = mask;
-	} else {
-		pio->PIO_PUDR = mask;
-	}
 
```

Comment:
Let's remove this empty line. The code will read better. Originally it wasn't there.